src/core/or/circuitpadding.c

```diff
@@ -0,0 +1,2221 @@
+
+  if (machine->conditions.requires_vanguards) {
+    const or_options_t *options = get_options();
+
+    /* Pinned middles are effectively vanguards */
+    if (!(options->HSLayer2Nodes || options->HSLayer3Nodes))
+      return 0;
+  }
+
+  if (!(circpad_circuit_state(circ) & machine->conditions.state_mask))
+    return 0;
+
+  if (circuit_get_cpath_opened_len(circ) < machine->conditions.min_hops)
+    return 0;
+
+  return 1;
+}
+
+/**
+ * Returns a minimized representation of the circuit state.
+ *
+ * The padding code only cares if the circuit is building,
+ * opened, or used for streams. This returns which one is true.
+ */
+static inline
+circpad_circuit_state_t
+circpad_circuit_state(origin_circuit_t *circ)
+{
+  if (circ->p_streams)
+    return CIRCPAD_CIRC_STREAMS;
+
+  /* We use has_opened to prevent cannibialized circs from flapping. */
+  if (circ->has_opened)
+    return CIRCPAD_CIRC_OPENED;
+
+  return CIRCPAD_CIRC_BUILDING;
+}
+
+/**
+ * Convert a normal circuit purpose into a bitmask that we can
+ * use for determining matching circuits.
+ */
+static inline
+circpad_purpose_mask_t
+circpad_circ_purpose_to_mask(uint8_t circ_purpose)
+{
+  /* Treat OR circ purposes as ignored. They should not be passed here*/
+  if (BUG(circ_purpose <= CIRCUIT_PURPOSE_OR_MAX_)) {
+    return 0;
+  }
+
+  /* Treat new client circuit purposes as "OMG ITS EVERYTHING".
+   * This also should not happen */
+  if (BUG(circ_purpose - CIRCUIT_PURPOSE_OR_MAX_ - 1 > 32)) {
+    return CIRCPAD_PURPOSE_ALL;
+  }
+
+  /* Convert the purpose to a bit position */
+  return 1 << (circ_purpose - CIRCUIT_PURPOSE_OR_MAX_ - 1);
+}
+
+/**
+ * Shut down any machines whose conditions no longer match
+ * the current circuit.
+ */
+static void
+circpad_shutdown_old_machines(origin_circuit_t *on_circ)
+{
+  circuit_t *circ = TO_CIRCUIT(on_circ);
+
+  for (int i = 0; i < CIRCPAD_MAX_MACHINES; i++) {
+    if (!circ->padding_machine[i])
+      continue;
+
+    if (!circpad_machine_conditions_met(on_circ,
+                                        circ->padding_machine[i])) {
+      // Clear machineinfo (frees timers)
+      circpad_circuit_machineinfo_free_idx(circ, i);
+      // Send padding negotiate stop
+      circpad_negotiate_padding(on_circ,
+                                circ->padding_machine[i]->machine_num,
+                                circ->padding_machine[i]->target_hopnum,
+                                CIRCPAD_COMMAND_STOP);
+    }
+  }
+}
+
+/**
+ * Negotiate new machines that would apply to this circuit.
+ */
+static void
+circpad_add_matching_machines(origin_circuit_t *on_circ)
+{
+  circuit_t *circ = TO_CIRCUIT(on_circ);
+
+#ifdef TOR_UNIT_TESTS
+  /* Tests don't have to init our padding machines */
+  if (!origin_padding_machines)
+    return;
+#endif
+
+  for (int i = 0; i < CIRCPAD_MAX_MACHINES; i++) {
+    if (circ->padding_info[i])
+      continue;
+
+    /* We have a free index */
+    SMARTLIST_FOREACH_REVERSE_BEGIN(origin_padding_machines,
```

Comment:
Why do we go reverse here?

In general, I count more than 10 for loops that loop over machines and another 5 smartlist lo ops that loop over machines. Some of those loops are nested loops...

Can we introduce some functions that abstract those loops into something more convenient and easier to understand? For example, we could have a function that  gets us the circ->padding_machine or the circpad_machine_t from the machine  type.

Reply:
We go in reverse so that when we add additional machine definitions in the consensus or torrc, those take precedence over older ones, and over any hardcoded machines.

Most of the loops are just over two padding machines indexes. I am not sure how to simplify them further. Though your comment did remind me that we need to check each index each time: 9d7e119aabcf1680df8aa9126b5ca9582176e999

This one nested loop is more efficient after we already have machines on the circuit than doing it the other way around.

Reply:
How about 370f33ff41 as an attempt to abstract away these loops? Hide away implementation details and make it more obvious what they do?

We have a similar pattern in `hs_service.c` (see `FOR_EACH_SERVICE_BEGIN` and `FOR_EACH_DESCRIPTOR_BEGIN`) to reduce the loop details around the codebase, and also to make it less likely we forget checks like the ones from 9d7e119.

Reply:
OK I took this and tweaked it a bit and did f7dc98f14be33873116fba08d1b0a70df5f34fdd